accepted/system-range/system-range.md

```diff
@@ -0,0 +1,246 @@
+
+        public Index(int value, bool fromEnd = false)
+        {
+            if (value < 0)
+                throw new ArgumentException("Index must not be negative.", nameof(value));
+
+            _value = fromEnd ? ~value : value;
+        }
+
+        public static implicit operator Index(int value) => new Index(value);
+    }
+
+    public readonly struct Range
+    {
+        public Index Start { get; }
+        public Index End { get; }
+
+        private Range(Index start, Index end)
+        {
+            Start = start;
+            End = end;
+        }
+
+        public static Range Create(Index start, Index end) => new Range(start, end);
+        public static Range FromStart(Index start) => new Range(start, new Index(0, fromEnd: true));
+        public static Range ToEnd(Index end) => new Range(new Index(0, fromEnd: false), end);
+        public static Range All() => new Range(new Index(0, fromEnd: false), new Index(0, fromEnd: true));
+    }
+```
+
+### Indexing with SZ-arrays
+
+SZ-arrays (i.e. `T[]` in C#) do not have an API that represents the indexer.
+Instead, they are specialized as IL opcodes to load and store elements in such
+an array. The indexer is part of the language semantics and thus doesn't have an
+API representation.
+
+For indexing using `Index` the compiler doesn't have to do much, thus the
+handling is done directly by the compiler, i.e. this code:
+
+```C#
+T[] array = GetArray(;)
+Index index = GetIndex();
+T element = array[index];
+```
+
+will be compiled down to the following code:
+
+```C#
+T[] array = GetArray(;)
+Index index = GetIndex();
+int $index = index.FromEnd
+                ? array.Length - index.Value
+                : index.Value;
+T element = array[$index];
+```
+
+For `Range` the code-gen would be a more a bit more involved, thus we propose
+to give the compiler a `RuntimeHelper` API to call, so this code:
+
+```C#
+T[] array = GetArray(;)
+Range range = GetRange();
+T[] elements = array[range];
+```
+
+will be compiled down to:
+
+```C#
+T[] array = GetArray(;)
+Range range = GetRange();
+T[] elements = RuntimeHelpers.GetArrayRange<T>(array, range);
+```
+
+We propose the following method:
+
+```C#
+namespace System.Runtime.CompilerServices
+{
+    public sealed class RuntimeHelpers
+    {
+        public static T[] GetArrayRange<T>(T[] array, Range range);
+    }
+}
+```
+
+### Companion APIs
+
+To decide which indexers we offer we followed the following principles:
+
+* Scalar indexers should return the element type
+* Range indexers should return the type they are on
+
+If developers whish to optimize for allocations they can convert the type to an
+instance of `Span<T>`, `ReadOnlySpan<T>`, `Memory<T>` or `ReadOnlyMemory<T>` by
+calling the appropriate API (e.g. `AsSpan()` or `AsMemory()`).
+
+We'll also provide overloads to common operations that take indexes and ranges,
+such as `String.Substring(int)`, `String.Substring(int, int)` and
+`AsSpan(int, int)`.
```

Comment:
Should the "such as" here be the existing overloads, or the new ones?  (I expected to see `String.Substring(Index)`, `String.Substring(Range)`, and `String.AsSpan(Range)`)